Source/Core/Common/MemArenaWin.cpp

```diff
@@ -1,71 +1,410 @@
-  hMemoryMapping = CreateFileMapping(INVALID_HANDLE_VALUE, nullptr, PAGE_READWRITE, 0,
-                                     static_cast<DWORD>(size), UTF8ToTStr(name).c_str());
+  m_memory_handle = CreateFileMapping(INVALID_HANDLE_VALUE, nullptr, PAGE_READWRITE, 0,
+                                      static_cast<DWORD>(size), UTF8ToTStr(name).c_str());
 }
 
 void MemArena::ReleaseSHMSegment()
 {
-  CloseHandle(hMemoryMapping);
-  hMemoryMapping = 0;
+  if (!m_memory_handle)
+    return;
+  CloseHandle(m_memory_handle);
+  m_memory_handle = nullptr;
 }
 
 void* MemArena::CreateView(s64 offset, size_t size)
 {
-  return MapInMemoryRegion(offset, size, nullptr);
+  return MapViewOfFileEx(m_memory_handle, FILE_MAP_ALL_ACCESS, 0, (DWORD)((u64)offset), size,
+                         nullptr);
 }
 
 void MemArena::ReleaseView(void* view, size_t size)
 {
-  UnmapFromMemoryRegion(view, size);
+  UnmapViewOfFile(view);
 }
 
 u8* MemArena::ReserveMemoryRegion(size_t memory_size)
 {
-  u8* base = static_cast<u8*>(VirtualAlloc(nullptr, memory_size, MEM_RESERVE, PAGE_READWRITE));
-  if (!base)
+  if (m_reserved_region)
   {
-    PanicAlertFmt("Failed to map enough memory space: {}", GetLastErrorString());
+    PanicAlertFmt("Tried to reserve a second memory region from the same MemArena.");
     return nullptr;
   }
-  VirtualFree(base, 0, MEM_RELEASE);
+
+  u8* base;
+  if (m_api_ms_win_core_memory_l1_1_6_handle)
+  {
+    base = static_cast<u8*>(static_cast<PVirtualAlloc2>(m_address_VirtualAlloc2)(
+        nullptr, nullptr, memory_size, MEM_RESERVE | MEM_RESERVE_PLACEHOLDER, PAGE_NOACCESS,
+        nullptr, 0));
+    if (base)
+    {
+      m_reserved_region = base;
+      m_regions.emplace_back(base, memory_size, false);
+    }
+    else
+    {
+      PanicAlertFmt("Failed to map enough memory space: {}", GetLastErrorString());
+    }
+  }
+  else
+  {
+    NOTICE_LOG_FMT(MEMMAP, "VirtualAlloc2 and/or MapViewFromFile3 unavailable. "
+                           "Falling back to legacy memory mapping.");
+    base = static_cast<u8*>(VirtualAlloc(nullptr, memory_size, MEM_RESERVE, PAGE_READWRITE));
+    if (base)
+      VirtualFree(base, 0, MEM_RELEASE);
+    else
+      PanicAlertFmt("Failed to find enough memory space: {}", GetLastErrorString());
+  }
+
   return base;
 }
 
 void MemArena::ReleaseMemoryRegion()
 {
+  if (m_api_ms_win_core_memory_l1_1_6_handle && m_reserved_region)
+  {
+    // user should have unmapped everything by this point, check if that's true and yell if not
+    // (it indicates a bug in the emulated memory mapping logic)
+    for (const auto& r : m_regions)
+    {
+      size_t mapped_region_count = 0;
+      if (r.m_is_mapped)
+        ++mapped_region_count;
+      if (mapped_region_count > 0)
+      {
+        PanicAlertFmt("Error while releasing fastmem region: {} regions are still mapped!",
+                      mapped_region_count);
+      }
+    }
```

Comment:
This logic is wrong; `mapped_region_count`'s declaration and the panic alert should be outside of the loop.  With the current logic, `mapped_region_count` will only ever be 0 or 1, and if multiple regions were not unmapped, you'll get multiple panic alerts.